### PR TITLE
- adds support for go snippets generation

### PR DESCRIPTION
**Pull request by baywet**
## Overview

Adds snippets generation for Go OData/Kiota

## Testing Instructions

```
POST /api/snippets?generation=openapi&lang=go
--BODY---
GET /v1.0/me HTTP/1.1
Host: graph.microsoft.com
Content-Type: application/json
```

Results in the new open api & kiota based generation with 

```Go
//THE GO SDK IS IN PREVIEW. NON-PRODUCTION USE ONLY
graphClient := msgraphsdk.NewGraphClient(requestAdapter);

result, err := graphClient.Me().Get(nil);
```

### CODE REVIEW

File: CodeSnippetsReflection.OpenAPI.Test/GoGeneratorTests.cs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    public class GoGeneratorTests {
+        private const string ServiceRootUrl = "https://graph.microsoft.com/v1.0";
+        private static OpenApiUrlTreeNode _v1TreeNode;
+        private static async Task<OpenApiUrlTreeNode> GetV1TreeNode() {
+            if(_v1TreeNode == null) {
+                _v1TreeNode = await SnippetModelTests.GetTreeNode("https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/v1.0/openapi.yaml");
+            }
+            return _v1TreeNode;
+        }
+        private readonly GoGenerator _generator = new();
```
Comment (thewahome) [inline CodeSnippetsReflection.OpenAPI.Test/GoGeneratorTests.cs:20]: Wasted chance to call it the GoGetter

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+					} else
+						payloadSB.AppendLine($"{indentManager.GetIndent()}{NormalizeQueryParameterName(queryParam)}: \"\",");
+				}
+				indentManager.Unindent();
+				payloadSB.AppendLine($"{indentManager.GetIndent()}}}");
+				return (payloadSB.ToString(), requestParametersVarName);
+			}
+			return (default, default);
+		}
+		private static Regex nestedStatementRegex = new(@"(\w+)(\([^)]+\))", RegexOptions.IgnoreCase);
```
Comment (andrueastman) [inline CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs:98]:
```suggestion
		private static Regex nestedStatementRegex = new(@"(\w+)(\([^)]+\))", RegexOptions.IgnoreCase | RegexOptions.Compiled);
```
I think we can still do this?

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    {
+        private const string clientVarName = "graphClient";
+		private const string clientVarType = "GraphServiceClient";
+		private const string httpCoreVarName = "requestAdapter";
+        public string GenerateCodeSnippet(SnippetModel snippetModel)
+        {
+            var indentManager = new IndentManager();
+			var snippetBuilder = new StringBuilder(
+									"//THE GO SDK IS IN PREVIEW. NON-PRODUCTION USE ONLY" + Environment.NewLine +
+									$"{clientVarName} := msgraphsdk.New{clientVarType}({httpCoreVarName});{Environment.NewLine}{Environment.NewLine}");
```
Comment (andrueastman) [inline CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs:23]:
```suggestion
									$"{clientVarName} := msgraphsdk.New{clientVarType}({httpCoreVarName}){Environment.NewLine}{Environment.NewLine}");
```
Do we really need the semicolon before the newlines? 
If not, I think we might be better off removing to be consistent with the rest of the generated code?

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+                                                .ToList();
+                                            var paramSet = string.Join(", ", parameters);
+											return x.Segment.Split('.').Last().ToFirstCharacterUpperCase() + $"({paramSet}).";
+                                        }
+										return x.Segment.ToFirstCharacterUpperCase() + "().";
+									})
+						.Aggregate((x, y) => $"{x}{y}")
+                        .Replace("().ById(", "ById(");
+		}
+		private static string GetMethodName(HttpMethod method) {
```
[the remark this answers is not in the document]
Comment (baywet) [reply CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs:277]: thanks, I'm not sure why I didn't go this way the first time around.... 🤦‍♂️

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,290 @@
+using System;
```
[the remark this answers is not in the document]
Comment (andrueastman) [reply CodeSnippetsReflection.OpenAPI/LanguageGenerators/GoGenerator.cs:1]: Resolved via https://github.com/microsoftgraph/microsoft-graph-devx-api/pull/749/commits/e93650e78acbdf160c137e47377033c06424c189

**Review by andrueastman — APPROVED** [review]

**Pull request merged**